File: app/src/main/java/com/nokhrin/corners/classical/GameOver.java
package com.nokhrin.corners.classical;


import com.nokhrin.corners.classical.view.ActivityClassic;

import static com.nokhrin.corners.resources.Constants.BLACK_CHECKER;
import static com.nokhrin.corners.resources.Constants.BOT_WIN;
import static com.nokhrin.corners.resources.Constants.PLAYER_WIN;
import static com.nokhrin.corners.resources.Constants.WHITE_CHECKER;
import static com.nokhrin.corners.resources.Constants.WIN_WIN;

public class GameOver {
    ActivityClassic activity;
    int[][] checkersPositions;

    public GameOver(ActivityClassic activity) {
        this.activity = activity;
    }

    public boolean isOver() {
        checkersPositions = activity.startGame.getCheckersPositions();

        int sizeOfField = checkersPositions.length;
        int countPointWhite = 0;
        int countPointBlack = 0;

        //sum white checker in home
        for (int i = 1; i < 4; i++) {
            for (int j = 5; j < sizeOfField; j++) {
                if (checkersPositions[i][j] == WHITE_CHECKER) {
                    countPointWhite++;
                }
            }
        }

        //sum black checker in home
        for (int i = 6; i < sizeOfField; i++) {
            for (int j = 1; j < 5; j++) {
                if (checkersPositions[i][j] == BLACK_CHECKER) {
                    countPointBlack++;
                }
            }
        }


        int countTargetPoint = activity.startGame.getCountTargetPoint();
        if (countPointWhite == countTargetPoint && countPointBlack == countTargetPoint) {
            activity.startGame.setWin(WIN_WIN);
            return true;
        } else if (countPointWhite == countTargetPoint) {
            activity.startGame.setWin(PLAYER_WIN);
            return true;
        } else if (countPointBlack == countTargetPoint) {
            activity.startGame.setWin(BOT_WIN);
            return true;
        } else {
            return false;
        }


    }
}

File: app/src/main/java/com/nokhrin/corners/levels/model/GameOver.kt
package com.nokhrin.corners.levels.model

import android.content.ContentValues
import android.util.Log
import com.nokhrin.corners.levels.database.LevelsDb.*
import com.nokhrin.corners.levels.database.ReadDb
import com.nokhrin.corners.levels.view.ActivityGameLevel
import com.nokhrin.corners.resources.Constants

class GameOver {
    private var countBronze = 0
    private var countSilver = 0
    private var countGold = 0
    private var activity: ActivityGameLevel? = null

    fun setActivity(activity: ActivityGameLevel) {
        this.activity = activity
        countBronze = activity.startGame.countBronze
        countSilver = activity.startGame.countSilver
        countGold = activity.startGame.countGold
    }
    val isOver: Boolean
        get() {
            var result = false
            var checkersPositions = activity?.startGame?.checkerPositions
            var marksPositions = activity?.startGame?.marksPositions
            var sizeOfField = checkersPositions?.size
            var countPointWhite = 0

            //sum white checker in home
            if (sizeOfField == null) sizeOfField = 0
            if(checkersPositions == null) checkersPositions = Array(1) { IntArray(1) }
            if(marksPositions == null) marksPositions = Array(1) { IntArray(1) }
            for (i in 1 until sizeOfField) {
                for (j in 1 until sizeOfField) {
                    if ((marksPositions[i][j] ==   Constants.TARGET_POINT_FOR_WHITE_CHECKER)
                        && (checkersPositions[i][j] == Constants.WOODMAN_CHECKER)
                    ) {
                        countPointWhite++
                    }
                }
            }
            Log.d(Constants.TAG, "isOver: countPointWhite = $countPointWhite")

            if (countPointWhite == activity?.startGame?.countTargetPoint) {
                activity?.startGame?.win = Constants.PLAYER_WIN
                //rate()
                result = true
            } else if (activity?.startGame?.countToMove == 0) {
                activity?.startGame?.win = Constants.BOT_WIN
                //rate()
                result = true
            }
            rate()
            return result
        }

    private fun getScore(): Int {
        val resultMove = activity?.startGame?.countToMove ?: return 0
        return when {
            resultMove >= countGold -> 3
            resultMove >= countSilver -> 2
            resultMove >= countBronze -> 1
            else -> 0
        }
    }

    private fun rate() {
        val score = getScore()
        val readDb = ReadDb()
        readDb.setLevelsDb(activity?.levelsDb)
        val processLevel = activity?.numberLevel?.let { readDb.progress[it] }
        if(processLevel != null) if(score <= processLevel) return

        val contentValues = ContentValues()
        contentValues.put(KEY_COUNT_STARS, score)
        contentValues.put(KEY_NUMBER_LEVEL, activity?.numberLevel)
        val dataBase = activity?.levelsDb?.writableDatabase

        dataBase?.update(TABLE_PROGRESS, contentValues, "$KEY_NUMBER_LEVEL= ?", arrayOf(
            activity?.numberLevel.toString()
        ))

        activity?.levelsDb?.close()
    }
}